contracts/GSDINFT.sol

```diff
@@ -0,0 +1,228 @@
+  function transferBorrowerAndCall(
+      uint256 _id,
+      address _receiver,
+      uint256 amount,
+      bytes calldata data
+  ) external override {
+    transferBorrower(_id, _receiver);
+
+    IGSDIBorrowerReceiver receiver_ = IGSDIBorrowerReceiver(_receiver);
+    receiver_.onBorrowerTransferred(msg.sender, amount, data);
+  }
+
+  /// @notice Mints a new GSDI in proposal to IGSDINFT. Locks the IGSDIWallet by setting IGSDINFT as the wallet's executor. Only callable by the current IGSDIWallet executor.
+  /// @param _maturity Timestamp when the GSDI matures and the lender may seize the wallet.
+  /// @param _faceValue Amount of currency borrower must pay to cover the GSDI.
+  /// @param _price Amount of currency the lender must pay the borrower to mint the GSDI.
+  /// @param _wallet Wallet containing collateral backing the GSDI.
+  /// @param _currency Token currency for the price and face value.
+  /// @param _borrower Address which will become the wallet executor after the GSDI is covered.
+  function propose(
+    uint256 _maturity,
+    uint256 _faceValue,
+    uint256 _price,
+    IGSDIWallet _wallet,
+    address _currency,
+    address _borrower
+  ) external override returns (uint256 id_) {
+    require(_wallet.executor() == msg.sender, "Wallet executor must be sender");
+
+    uint256 tokenId = _tokenIdTracker.current(); 
+    id_ = abi.encodePacked(tokenId >> 96, chainId << 160).toUint256(0);
+
+    metadata[id_] = NFTMetadata(_maturity, _faceValue, _price, _wallet, _currency, _borrower, true);
+
+    // Locks the IGSDIWallet by setting IGSDINFT as the wallet's executor.
+    _wallet.setExecutor(address(this));
+
+    _mint(_borrower, id_);
+    _tokenIdTracker.increment();
+
+    emit Propose(id_, tokenId, _borrower, address(this));
+  }
+
+  ///@notice Cancels the GSDI proposal. Sender must be borrower. GSDI must currently be in proposal. GSDI must be on the current chain. Burns the GSDI.
+  /// @param _id GSDI to cancel.
+  function cancel(uint256 _id) external override mustBeInProposal(_id) onlyValidChainId(_id) {
+    require(msg.sender == metadata[_id].borrower, "Sender must be borrower");
+    burnProposal(_id);
+    
+    emit Cancel(_id);
+  }
+
+  /// @notice Sends the GSDI to sender. The IGSDIWallet must be in proposal. GSDI must be on the current chain. Transfers price in currency from sender to borrower. Removes GSDI from proposal.
+  /// @notice If governance has enabled the 0.3% fee, the sender must also transfer the fee to the governance's treasury.
+  /// @dev Sender must approve the contract to transfer price of currency before call.
+  /// @param _id GSDI to purchase.
+  function purchase(uint256 _id) external override mustBeInProposal(_id) onlyValidChainId(_id) {
+    uint256 price = metadata[_id].price;
+    IERC20 token = IERC20(metadata[_id].currency);
+
+    uint256 _before = token.balanceOf(address(this));
+    token.safeTransferFrom(msg.sender, address(this), price);
```

Comment:
The dai token contract (and many other erc20 contracts) do not support safeTransfer/safeTransferFrom. Instead, call rever(token.transfer(...),"GSDINFT: Transfer failed.") both here and other places safetransfer/safetransferfrom are used.